libs/pageserver_api/src/record.rs

```diff
@@ -0,0 +1,113 @@
+//! This module defines the WAL record format used within the pageserver.
+
+use bytes::Bytes;
+use postgres_ffi::record::{describe_postgres_wal_record, MultiXactMember};
+use postgres_ffi::{MultiXactId, MultiXactOffset, TimestampTz, TransactionId};
+use serde::{Deserialize, Serialize};
+use utils::bin_ser::DeserializeError;
+
+/// Each update to a page is represented by a NeonWalRecord. It can be a wrapper
+/// around a PostgreSQL WAL record, or a custom neon-specific "record".
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum NeonWalRecord {
```

Comment:
> It doesn't feel right to have this in pageserver_api. What API is this part of? Most of the stuff in this module are part of the pageserver HTTP, and this is surely not that.
I presume it's for the safekeeper -> pageserver connection. But we don't have others parts of that API in this module. It's just a stream of WAL at the moment. I guess we're evolving that API to something more complicated; what will the API look like when it's finished?

I broadly feel the same way. Here's what my thought process was:

Safekeepers will send interpreted WAL records to pageservers ([InterpretedWalRecord](https://github.com/neondatabase/neon/blob/7c4807e4b985e3fa40533459c61eda06022a3fe3/libs/wal_decoder/src/models.rs#L22-L37) from a [draft of the follow-up](https://github.com/neondatabase/neon/pull/9524)). These currently include `pageserver_api::value::Value` instances, but eventually these will be replaced with a raw buffer which can be ingested directly by writing to the ephemeral file. I needed the `Value` type to live in `pageserver_api` to enable this arrangement, and, transitively, `NeonWalRecord` also had to move.

Looking back at this PR, I only need the `NeonWalRecord::Postgres` variant for the decoder module. All the other variants will only be generated by the pageserver in `walingest`. I could define a new `RawValue` with just two variants `Image` and `Postgres`, and move `Value` and `NeonWalRecord` back into `pageserver`. I think this might work out nicer. Will try.

Reply:
I ended up moving `NeonWalRecord` and `Value` back into the pageserver crate. I think it ultimately makes sense for them to live there.

Reply:
And then I realised that I need access to the `Value` type such that the safekeeper will be able to serialize them into a buffer ready for dumping into the ephemeral file. Moved them back ...

Reply:
Will resolve and merge to unblock the next PR in the series. Let's move to a Slack thread if needed.